src/include/misc_inline.h

```diff
@@ -315,6 +315,36 @@ __wt_read_shared_double(double *to_read)
     return (*to_read);
 }
 
+/*
+ * __wt_set_shared_maximum64 --
+ *     Given an integer in memory overwrite it with a new value if it is larger.
+ */
+static WT_INLINE void
+__wt_set_shared_maximum64(uint64_t *shared_value, uint64_t new_val)
+{
+    uint64_t current;
+
+    current = *shared_value;
+    while (current < new_val) {
```

Comment:
If this exits because `!(current < new_value)` you will not get the expected barrier semantics out of this primitive. As much as it pains me to say it, you need to so something like this:

```c
while (true) {
  if (current >= new_val) {
    fetch_add(shared_value, 0); // no-op SeqCst RMW
    break;
  }
  if (compare_exchange(...))
    break;
}
```

That will ensure that this always behaves like a barrier. Note that the `ldumaxal` instruction is defined in such a way that this isn't necesary there, just always use that instruction. Also I'm pretty sure that the x86 `lock; cmpxchg;` instruction also has strong semantics even on failure, so if you want to go to ASM there, you would only need the barrier if you would be bailing before the first cmpxchg (but if you do this we should double check the spec.)

See section 5 of the [proposal](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p0493r5.pdf) to standardize `fetch_max` for a bit of discussion. In a mailing list discussion I tried to argue against this, but I think they convinced me that it is at least observable in the C++ memory model, even if only under very contrived examples. Given that all other RMWs are defined as always performing both the read and write (even when no-ops like `fetch_add(0)`, it seems consistent to use the same definition for min/max. Ditto in the WT model where RMWs are defined as "full barriers".

Reply:
The two usages we have for this act as a `set-and-forget` model, which I don't think requires release semantics, however I could see usages in the future that may want release semantics. This touches on a not well defined part of WiredTiger where we don't specify what memory ordering semantics a function call gives you. If we were to add this I believe we would mandate a minimum ordering of release and not leave it up to the user to specify, this would then be added to the function comment. 

I will need to think about this more. The document has this to say:
"if the user requested memory order is not a release, then store is not required" to confirm the three memory orders that aren't a release are relaxed, acquire, and consume? Where seq_cst is an upgrade on a release.

cc: @ajmorton

Reply:
So I've been thinking about the release semantic part of this, I can't see why we'd need it at least given the two usages in the code that we have now. I do understand that if we don't need to store anything and don't enter the loop we lose any ordering semantics, so we could strengthen it to release semantics (loads and stores don't cross).

I was discussing this with @dgomezferro, and we even considered a relaxed CAS which seemed to provide sufficient semantics given this [cppmem](http://svr-pes20-cppmem.cl.cam.ac.uk/cppmem/index.html) test:
```
int main() {
  atomic_int x = 2;
  {{{ 
   {
   int y = x.load(mo_relaxed).readsvalue(2);
   y = cas_strong_explicit(x, y, 5, mo_relaxed, mo_relaxed);
}
  ||| 
{
   int z = x.load(mo_relaxed);
   z = cas_strong_explicit(x, z, 6, mo_relaxed, mo_relaxed);
   z.readsvalue(6);
}
  }}};
  x.load().readsvalue(5);
  return 0; }
```

Obviously to be consistent we will use a `SEQ_CST` CAS but it did raise the question of what ordering do we need here. I will continue to think on this and see if I can describe a scenario in WT that would be end badly if we didn't at least have release semantics on call to this function.

Reply:
This value is read in two additional places:
bt_sync.c:281
```
                if (mod != NULL && mod->rec_max_txn > btree->rec_max_txn)
                    btree->rec_max_txn = mod->rec_max_txn;
                if (mod != NULL && btree->rec_max_timestamp < mod->rec_max_timestamp)
                    btree->rec_max_timestamp = mod->rec_max_timestamp;
```
bt_sync.c:370:
```
        if (!btree->modified && !F_ISSET(conn, WT_CONN_RECOVERING | WT_CONN_CLOSING_CHECKPOINT) &&
          (btree->rec_max_txn >= txn->snapshot_data.snap_min ||
            (conn->txn_global.checkpoint_timestamp != conn->txn_global.last_ckpt_timestamp &&
              btree->rec_max_timestamp > conn->txn_global.checkpoint_timestamp)))
            __wt_tree_modify_set(session);
```
These are all within the context of the `btree->flush_lock` . Now it is evident that the reconciliation code does not take this lock. So there is still the potential to race here. With `rec_max_txn` an `rec_max_timestamp` being written in that order it would be important to read them in that order which does define a `release` requirement on at least `rec_max_txn` and acquire semantics for `rec_max_timestamp`.  

If we write A->B we should read B->A which the code currently does not do. The easiest way to fix this would be to lock in the reconciliation code but that could lead to deadlocks. So for now I will do the following:
1. Guarantee release semantics for the function call into `__wt_set_shared_maximum_u64`
2. Fix bt_sync.c:281 to utilize the CAS functionality
3. Modify the order of reads and perform an acquire load at bt_sync.c:370

Reply:
I was thinking of message passing when it came to ordering these writes, I still can't define a valid order between them. This would be much easier if there was a lock to utilize here.

Reply:
I don't think there's any way to do this obviously correctly other than using a lock. In theory any key value pair in the tree with txnid or timestamp greater than the btree->rec_max_txn and btree->rec_max_timestamp should trigger the `__wt_tree_modify_set`.

Which means even if we locked around writing to these fields in reconciliation there's nothing that would prevent another write happening which would then move these values. I think without a fully fledged review of this code and the related code paths I won't have a reliable solution here. I am probably going to kick off a google doc.

Reply:
@luke-pearson When I tried the cppmem test posted above in cppm I got: 112 executions; no consistent.
I'm curious as to which model you used (none worked for me) and how you drew your conclusion. Thanks.

Reply:
That is correct there is no consistent execution that will end with reading the value 5 from x at the end of the execution. In thread 1 we attempt to swap 2 with 5, that is the state that `x.readsvalue(2)` creates. We don't check this for success as we don't have another readsvalue check for that thread. Thread 2 attempts to swap in 6 for whatever value exists. Given thread 1 reads 2 we have two scenarios:

1. Thread 1 CAS Success: The value 5 was swapped in but we didn't read that value at the end of execution, this means that 2 was overwritten by 6
2. Thread 1 CAS Failure: Nothing was swapped in by thread 1 as the thread 2 CAS happened after we read 2, therefore 6 must exist at the end.

I'm sure you get the point but yes there's no way to reach an end state which has a value of 5 for x.

Reply:
Interesting. I think a constraint is missing, the following gives a consistent execution:
```
int main() {
  atomic_int x = 2;
  {{{ 
   {
   int y = x.load(mo_relaxed); // X
   y = cas_strong_explicit(x, y, 5, mo_relaxed, mo_relaxed);
   y.readsvalue(5);
}
  ||| 
{
   int z = x.load(mo_relaxed).readsvalue(2);
   z = cas_strong_explicit(x, z, 6, mo_relaxed, mo_relaxed);
   z.readsvalue(6);
}
  }}};
  x.load().readsvalue(5);
  return 0; }
```
But adding `readsvalue(2);` at position `X` forces a data race  on x.

Reply:
@marcbutler as far as I know cpp mem doesn't support ifs/whiles, that's why I simulated the execution I was interested in with `.readsvalue()`s. If you remove `readsvalue(2);` from the first thread then it will read 6 and overwrite it with 5, which is something that wouldn't happen with Luke's C implementation (we'd see that 5 < 6 and bail out).